### ISSUE TITLE
Unachieved Expected Performance

### ISSUE THREAD

**Issue by 11z000i** [snap-stanford/roland#10]
Hi,
I have used the code to run some experiments, but it seems that I haven't achieved the expected performance. Below are the commad I ran and the results I obtained:
`python3 main.py --cfg './run/replication_configs/table2/BitcoinOTC_GRU.yaml' --repeat 1 --override_data_dir '/data/shot_data' --override_remark 'roland_example'`
![image](https://github.com/snap-stanford/roland/assets/108251027/26542202-ad1b-4487-ba72-7a7c0e137e6c)
`python3 main.py --cfg '/workspace/roland-master_v/run/replication_configs/table2/BitcoinAlpha_GRU.yaml' --repeat 1 --override_data_dir '/data/shot_data' --override_remark 'roland_example'`
![image](https://github.com/snap-stanford/roland/assets/108251027/08099c81-6b19-4930-b73a-b8b874861846)

The MRR here appears to be somewhat different from what was reported in the paper.

If this is because I made a mistake somewhere , or if my understanding of the experimental results is incorrect?

**Comment by TianyuDu** [comment]
Hey,

Thanks for reaching out! My apologies that I do not have access to the exact computing environment we were using to do these experiments. I will try my best and look into this issue shortly. Meanwhile, all Tensorboard records from our hyper-parameter searching in experiments were uploaded [here](https://github.com/snap-stanford/roland/tree/master/roland_raw_records) in this repository.

I looked into the script generating Table 2, the best configuration for GRU on BitCoinAlpha corresponds to the following configuration: `fixed_split\all_fixed_split_v11_grid_all_ours_fixed_split_v11\all_fixed_split_v11-model=residual_edge_conv-name=bitcoinalpha.csv-update=gru-mp=2-bn=True-msgdir=both-lr=0.3-post_mp=2\{1, 2, 3}`, where the suffix `{1, 2, 3}` corresponds to three random seeds.

The best configuration for GRU on BitCoinOTC was from `fixed_split\all_fixed_split_v11_grid_all_ours_fixed_split_v11\all_fixed_split_v11-model=residual_edge_conv-name=bitcoinotc.csv-update=gru-mp=4-bn=True-msgdir=both-lr=0.3-post_mp=2\{1, 2, 3}`

I have looked into the Tensorboard plot of these six experiments (2 datasets, 3 random seeds each) and the log result seems to match our numbers in Table 2 (the test period starts at around snapshot 108 in both datasets)

<img width="1946" alt="image" src="https://github.com/snap-stanford/roland/assets/22979098/f4c93e2d-c68f-4d18-ac48-fd957acda0b5">

Please feel free to fetch these tensorboard experiment logs. I understand that there are lots of tensorboard logs in that directory, I have copy-and-pasted the six relevant experiment logs for you [select_6_experiments.zip](https://github.com/snap-stanford/roland/files/13259353/select.zip).

It's worth mentioning that PyTorch geometric evolved extremely fast these years and it's internal might changed. That would be a possible source of inconsistent results.

**Comment by TianyuDu** [comment]
Hey,
Sorry for the delayed reply; I spent a couple of hours looking into the issue by comparing the script I used for hyperparameter tuning and the replication scripts. The MRRs were incorrectly reported due to a typo I made while preparing for replication scripts, which has been fixed in this [commit](https://github.com/snap-stanford/roland/commit/609f858d04a23a9f66e7578e8becfc2801797836) I just made. I truly appreciate your comment so that I could spot the bug. The fixed method needs to run on CPU because it's less memory efficient (see details below), the script would take longer to run, typically < 5 min for each seed.

After the modification, I tried three random seeds on both the OTC and Alpha datasets, and here are MRRs I got:
* Bitcoin OTC: `0.24904646499944386, 0.24067379549618445, 0.17730400449575315`
* Bitcoin Alpha: `0.27158289165310917, 0.30106525715889887, 0.2877604731018733`


**Details**:
The computational of MRR requires two sets of edges: existing edges (positive edges) versus non-existing edges (negative edges). The MRR quantifies if the model will assign a high rank to existing edges compared to non-existing edges. In a sparse network, the set of existing edges is typically small, but the set of negative edges could be super large.

There are two natural ways to compute MRR: we can either use (1) all negative edges, which only works on small-size networks due to memory constraints or (2) sample a fixed amount of negative edges (say `1000 x num_positive_edges`).

In Table 2 we used method (1) because these graphs are small and we need to compare results to other papers. We used all negative edges to achieve consistency (as mentioned in the table caption: our experiments ensure the same snapshot frequency, set of test snapshots, and **method to compute MRRs**, etc).

In other tables, as we mentioned in the paper, because these graphs are potentially large, we sampled `1000 x num_positive_edges` negative edges for MRR computation. We used method (2) in these tables.

In the replication scripts, I mistakenly applied method (2) by calling `train_utils.report_rank_based_eval()` (as you can see in the commit, it takes a `num_neg_per_node` arg, which is the 1,000 multiplier mentioned above) to Table 2, we should use `train_utils.report_baseline_MRR()` instead.

**Comment by 11z000i** [comment]
Hi,
I run the updated codes and get the expected results. Thanks for your great work!

In addition, there is a minor issue: the configuration file `fixed_split\all_fixed_split_v11_grid_all_ours_fixed_split_v11\all_fixed_split_v11-model=residual_edge_conv-name=bitcoinalpha.csv-update=gru-mp=2-bn=True-msgdir=both-lr=0.3-post_mp=2\{1, 2, 3}` has 'dataset.edge_dim' set to 16, which cause the error below:
![image](https://github.com/snap-stanford/roland/assets/108251027/1f105ef0-52c8-4336-97ff-605b5a020e0f)

Is there something wrong with the file or the dataset?

**Comment by TianyuDu** [comment]
The `config.yaml` file saved after model training (the one you found in `fixed_split\all_fixed_split_v11_grid_all_ours_fixed_split_v11\all_fixed_split_v11-model=residual_edge_conv-name=bitcoinalpha.csv-update=gru-mp=2-bn=True-msgdir=both-lr=0.3-post_mp=2\{1, 2, 3}`) is indeed different from the input one because the training script wrote extra information such as `dataset.edge_dim` to the configuration. Could you please use `https://github.com/snap-stanford/roland/blob/master/run/replication_configs/clean_yaml.py` script to clean that yaml configuration and try again?

**Comment by 11z000i** [comment]
It solves the problem. Thanks for you explanation! :)
